server/service/osquery_utils/queries.go

```diff
@@ -628,7 +628,27 @@ var extraDetailQueries = map[string]DetailQuery{
 		// osquery table on darwin and linux, it is always present.
 	},
 	"disk_encryption_windows": {
-		Query:            `SELECT 1 FROM bitlocker_info WHERE drive_letter = 'C:' AND protection_status = 1;`,
+		Query: `
+WITH encrypted(enabled) AS (
+  SELECT CASE WHEN (
+    -- In some versions of Windows (server?), bitlocker is optional.
+    -- When bitlocker isn't installed, the bitlocker_info table query
+    -- will crash and not return any results, so we have to be careful
+    -- about when we query it.
+    SELECT 1 FROM windows_optional_features WHERE name = 'BitLocker'
+  ) THEN (
+    SELECT CASE WHEN (
+      -- If it's optional, and it's enabled
+      SELECT 1 FROM windows_optional_features WHERE name = 'BitLocker' AND state = 1
+    ) THEN (
+      -- Get bitlocker properties
+      SELECT 1 FROM bitlocker_info WHERE drive_letter = 'C:' AND protection_status = 1
+    ) END
+  ) ELSE (
+    -- Bitlocker is not optional, and should be on the device
+    SELECT 1 FROM bitlocker_info WHERE drive_letter = 'C:' AND protection_status = 1
+  ) END
+  ) SELECT enabled FROM encrypted WHERE enabled IS NOT NULL;`,
```

Comment:
awesome query! have you tried using a [discovery query](https://osquery.readthedocs.io/en/stable/deployment/remote/#discovery-queries-on-distributed-queries)? 

> A distributed discovery query controls whether or not a distributed query will be executed on a host.
> 
> - If a distributed query has no corresponding discovery query, then it is always executed on the host.
> - If a discovery query returns one or more results, then its corresponding distributed query will be executed on the host.
> - If a discovery query returns no results, then its corresponding distributed query will not be executed on the host.

In our system, you can define a `DiscoveryQuery` as part of the struct:

https://github.com/fleetdm/fleet/blob/a24f2e390e8c3894099e4aa6179664d98648d7d5/server/service/osquery_utils/queries.go#L36-L38

sorry if you already tried this! if it didn't work, could you add a comment or comment here why for future reference?

Reply:
I can rewrite it to use two discovery queries if you think it makes more sense!

Reply:
I would need two queries then, one for systems where bitlocker is optional and one for systems where bitlocker is built-in. I actually hadn't considered using a discovery query when I was writing this 😅. I'll try doing a quick rewrite and see if it looks any cleaner!

Reply:
@dantecatalfamo two queries would be super confusing, so if we can't do it in a single query please don't bother. This is fine as-is.

Just for fun, I was thinking something on the lines of (quick and dirty, haven't tested)

```sql
SELECT 1
WHERE NOT EXISTS (SELECT 1 FROM windows_optional_features WHERE name = 'BitLocker')
AND EXISTS (SELECT 1 FROM windows_optional_features WHERE state = 1);
```